active/0000-stateful-helpers.md

```diff
@@ -0,0 +1,322 @@
+
+Upon teardown:
+
+* The helper is destroyed, calling the `destroy` method.
+
+### Returning a value
+
+The return value of stateful and function helpers should be the return value
+of the HTMLBars subexpression they are powering. For example, given a helper:
+
+```js
+// app/helpers/full-name.js
+import Ember from "ember";
+
+// Note that what is called "options" in other docs is
+// called "blocks" here. This terminology should become the
+// documented norm.
+export default function full-name(params, hash, blocks) {
+  return params.join(' ');
+}
+```
+
+Or its equivilent stateful version, the following are effectively the same:
+
+```hbs
+<div data-name={{full-name "Fenton" "Crackshell"}}></div>
+<div data-name={{"Fenton Crackshell"}}></div>
+```
+
+```hbs
+{{my-component name=(full-name "Magica" "De Spell")}}
+{{my-component name="Magica De Spell"}}
+```
+
+```hbs
+<p>{{full-name "Bentina" "Beakley"}}</p>
+<p>{{"Bentina Beakley"}}</p>
+```
+
+An exclusion to this pattern is the following form:
+
+```hbs
+<div {{full-name "Webbigail" "Vanderquack"}}></div>
+```
+
+This is a legacy form of mustache usage. In 2.0 both forms of helpers should
+throw an exception when used in this manner.
+
+### Block form
+
+Helpers can optionally take a template and inverse block. To render a passed
+block, they must yield it. For example, an `if-is-yes` helper might be
+written like so:
+
+```js
+// app/helpers/custom-if.js
+import Ember from "ember";
+
+// Usage: {{if-is-yes 'no'}}not rendered{{else}}rendered!{{/if-is-yes}}
+export default function ifIsYes(params, hash, blocks) {
+  if (params[0] === 'yes') {
+    blocks.template.yield();
+  } else {
+    blocks.inverse.yield();
+  }
+};
+```
+
+Porting this usage to a stateful helper only changes the boilerplate:
+
+```js
+// app/helpers/custom-if.js
+import Ember from "ember";
+
+// Usage: {{if-is-yes 'no'}}not rendered{{else}}rendered!{{/if-is-yes}}
+export default Ember.Helper.extend({
+  compute(params, hash, blocks) {
+    if (params[0]) {
+      blocks.template.yield();
+    } else {
+      blocks.inverse.yield();
+    }
+  }
+});
+```
+
+### Consuming services and recompute
+
+Stateful helpers are a valid target for service injection. For example:
+
+```js
+// app/helpers/current-user-name.js
+import Ember from "ember";
+
+export default Ember.Helper.extend({
+  // Same API as components:
+  session: Ember.inject.service(),
+  compute() {
+    return this.get('session.currentUser.name');
+  }
+});
+```
+
+However consuming a property from a service does not bind the data being
+displayed to that property. After `{{current-user-name}}` has been computed
+and rendered, it will never be invalidated.
+
+For this reason, stateful helpers are granted some control over their
+computation lifecycle. A stateful helper will recompute when:
+
+* A value passed via the template changes (`params` or `hash`)
+* The `recompute` method is called
+* Any key in the `recomputeObserves` array changes
```

Comment:
I don't think we should add observation sugar right now. We can add sugar later (I personally think events / observables are going to be a more common pattern from services), and in the interim we will get more comfortable with even more ergonomic decorator approaches.